File: 139-word-break/word-break.cpp
class Solution {
public:
    bool solve(string s, int index, vector<string> &words, vector<int> &dp){
        if (index < 0) return true;
        if (dp[index] != -1) return dp[index] == 1;

        for(auto w:words){
            int len = w.length();

            if(index - len + 1 < 0) continue;

            if(s.substr(index - len + 1, len) == w && solve(s, index - len, words, dp)){
                return dp[index] = 1;
            }
        }

        return dp[index] = 0;
    }
    bool wordBreak(string s, vector<string>& words) {
        vector<int> dp(s.length(), -1);
        return solve(s, s.length()-1, words, dp);
    }
};

File: 139-word-break/word-break.go
func solve(s string, wordDict []string, idx int, dp []int) bool {
    if idx >= len(s) {
        return true
    }

    if dp[idx] != -1 {
        return dp[idx] == 1
    }

    dp[idx] = 0

    for _, str := range wordDict {
        length := len(str)

        if idx+length > len(s) {
            continue
        }

        if (s[idx:idx+length] == str) && solve(s, wordDict, idx+length, dp) {
            dp[idx] = 1
        }
    }

    return dp[idx] == 1
}

func wordBreak(s string, wordDict []string) bool {
    dp := initDP(len(s))
    return solve(s, wordDict, 0, dp)
}

func initDP(n int) (dp []int) {
    dp = make([]int, n+1)
    for i := range dp {
        dp[i] = -1
    }

    return dp
}